src/write/xcoff.rs

```diff
@@ -0,0 +1,496 @@
+                        strtab.get_offset(symbol_offsets[index].str_id.unwrap()) as u32,
+                    ),
+                    n_scnum: I16::new(BE, section_number),
+                    n_type: U16::new(BE, sym_type),
+                    n_sclass: storage_class,
+                    n_numaux: aux_num,
+                };
+                buffer.write(&xcoff_sym);
+            } else {
+                let mut sym_name = [0; 8];
+                let name = &symbol.name[..];
+                if name.len() <= 8 {
+                    sym_name[..name.len()].copy_from_slice(name);
+                } else {
+                    let str_offset = strtab.get_offset(symbol_offsets[index].str_id.unwrap());
+                    sym_name[4..8].copy_from_slice(&u32::to_be_bytes(str_offset as u32));
+                }
+                let xcoff_sym = xcoff::Symbol32 {
+                    n_name: sym_name,
+                    n_value: U32::new(BE, symbol.value as u32),
+                    n_scnum: I16::new(BE, section_number),
+                    n_type: U16::new(BE, sym_type),
+                    n_sclass: storage_class,
+                    n_numaux: aux_num,
+                };
+                buffer.write(&xcoff_sym);
+            }
+            // Generate a csect auxiliary entry for C_EXT, C_WEAKEXT, and C_HIDEXT Symbols.
+            if storage_class == xcoff::C_EXT
+                || storage_class == xcoff::C_WEAKEXT
+                || storage_class == xcoff::C_HIDEXT
+            {
+                debug_assert_eq!(aux_num, 1);
+                let (xsmtype, xsmclas) = if let Some(section) = section {
+                    match section.kind {
+                        SectionKind::Data => (xcoff::XTY_SD, xcoff::XMC_RW),
+                        SectionKind::Tls => (xcoff::XTY_SD, xcoff::XMC_TL),
+                        SectionKind::UninitializedData => (xcoff::XTY_CM, xcoff::XMC_BS),
+                        SectionKind::UninitializedTls => (xcoff::XTY_CM, xcoff::XMC_UL),
+                        SectionKind::ReadOnlyData | SectionKind::ReadOnlyString => {
+                            (xcoff::XTY_SD, xcoff::XMC_RO)
+                        }
+                        SectionKind::Text | SectionKind::OtherString => {
+                            if symbol.kind == SymbolKind::Label {
+                                (xcoff::XTY_LD, xcoff::XMC_PR)
+                            } else {
+                                (xcoff::XTY_SD, xcoff::XMC_PR)
+                            }
+                        }
+                        _ => {
+                            return Err(Error(format!(
+                                "unimplemented section `{}` kind {:?}",
+                                section.name().unwrap_or(""),
+                                section.kind
+                            )));
+                        }
+                    }
+                } else {
+                    return Err(Error(format!(
```

Comment:
There is situation we have a symbol with kind `SymbolSection::Undefined`, we will set `n_scnum` to `N_UNDEF` in line 378. This situation should be valid in XCOFF's view.